pyk/src/pyk/kast.py

```diff
@@ -1105,6 +1137,118 @@ def let_att(self, att: KAtt) -> 'KDefinition':
         return self.let(att=att)
 
 
+@dataclass(frozen=True)
+class Subst(Mapping[str, KInner]):
+    _subst: FrozenDict[str, KInner]
+
+    def __init__(self, subst: Mapping[str, KInner] = {}):
+        object.__setattr__(self, '_subst', FrozenDict(subst))
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._subst)
+
+    def __len__(self) -> int:
+        return len(self._subst)
+
+    def __getitem__(self, key: str) -> KInner:
+        return self._subst[key]
+
+    def __mul__(self, other: 'Subst') -> 'Subst':
+        return self.compose(other)
+
+    def __call__(self, term: KInner) -> KInner:
+        return self.apply(term)
+
+    def minimize(self) -> 'Subst':
+        return Subst({k: v for k, v in self.items() if v != KVariable(k)})
+
+    def compose(self, other: 'Subst') -> 'Subst':
+        from_other = ((k, self(v)) for k, v in other.items())
+        from_self = ((k, v) for k, v in self.items() if k not in other)
+        return Subst(dict(chain(from_other, from_self)))
+
+    def union(self, other: 'Subst') -> Optional['Subst']:
+        subst = dict(self)
+        for v in other:
+            if v in subst and subst[v] != other[v]:
+                return None
+            subst[v] = other[v]
+        return Subst(subst)
+
+    def apply(self, term: KInner) -> KInner:
+        def replace(term):
+            if type(term) is KVariable and term.name in self:
+                return self[term.name]
+            return term
+
+        return bottom_up(replace, term)
+
+    def unapply(self, term: KInner) -> KInner:
+        new_term = term
+        for var_name in self:
+            lhs = self[var_name]
+            rhs = KVariable(var_name)
+            new_term = KRewrite(lhs, rhs).replace(new_term)
+        return new_term
+
+
+def match(pattern: KInner, term: KInner) -> Optional[Subst]:
+    """Perform syntactic pattern matching and return the substitution.
+
+    -   Input: a pattern and a term.
+    -   Output: substitution instantiating the pattern to the term.
+    """
+
+    # TODO simplify
+    def _match(pattern: KInner, term: KInner) -> Optional[Dict[str, KInner]]:
+
+        def combine_dicts(dict1: Optional[Mapping[K, V]], dict2: Optional[Mapping[K, V]]) -> Optional[Dict[K, V]]:
```

Comment:
Now that we have `Subst` supporting a proper `_.union(_)` operator, maybe `_match(...)` can be made to return an `Optional[Subst]` instead of an `Optional[Dict[str, KInner]]`, and we can get rid of `combine_dicts` altogether?

Reply:
Done.